File: go/pkg/dns/dns.go
// Package dns implements the v1.4 DNS TXT cross-verification feature.
//
// A tool provider MAY publish a TXT record at _schemapin.{domain} containing
// the public-key fingerprint advertised in .well-known/schemapin.json. When
// present, clients use it as a second-channel verification: the DNS
// credential chain is independent of the HTTPS hosting credential chain, so
// compromising one does not compromise the other.
//
// TXT record format:
//
//	_schemapin.example.com. IN TXT "v=schemapin1; kid=acme-2026-01; fp=sha256:a1b2c3..."
//
// Fields:
//   - v   -- version tag (schemapin1); required
//   - fp  -- key fingerprint (sha256:<hex>); required, lowercase hex
//   - kid -- optional key id, used for disambiguating multi-key endpoints
//
// Verification semantics:
//   - Absent record           -> no effect (DNS TXT is optional)
//   - Present and matching    -> confidence boost (no warning emitted)
//   - Present and mismatching -> hard failure with verification.ErrDomainMismatch
//
// Use ParseTxtRecord to parse a raw TXT string and VerifyDnsMatch to
// cross-check it against a discovery document. FetchDnsTxt performs the
// actual lookup using the Go stdlib resolver -- no external dependency.
package dns

import (
	"context"
	"errors"
	"fmt"
	"net"
	"strings"

	"github.com/ThirdKeyAi/schemapin/go/pkg/crypto"
	"github.com/ThirdKeyAi/schemapin/go/pkg/discovery"
)

// DnsTxtRecord is a parsed _schemapin.{domain} TXT record.
type DnsTxtRecord struct {
	// Version is the protocol tag; currently always "schemapin1".
	Version string
	// Kid is the optional key id; empty when not present.
	Kid string
	// Fingerprint is the lowercase fingerprint string, including the
	// "sha256:" prefix. Always lower-case after parsing.
	Fingerprint string
}

// ParseTxtRecord parses a raw TXT record value such as
// "v=schemapin1; kid=acme-2026-01; fp=sha256:abcd...".
//
// Whitespace around ';' and '=' is tolerated. Field order is not significant.
// Unknown fields are ignored for forward compatibility. Returns an error if
// the record is missing the required 'v' or 'fp' fields, has an unsupported
// version, has a fingerprint without the "sha256:" prefix, or contains a
// field without an '=' separator.
func ParseTxtRecord(value string) (*DnsTxtRecord, error) {
	var (
		version string
		kid     string
		fp      string
		hasV    bool
		hasFp   bool
	)

	for _, raw := range strings.Split(value, ";") {
		part := strings.TrimSpace(raw)
		if part == "" {
			continue
		}
		idx := strings.Index(part, "=")
		if idx < 0 {
			return nil, fmt.Errorf("DNS TXT field missing '=': %s", part)
		}
		k := strings.ToLower(strings.TrimSpace(part[:idx]))
		v := strings.TrimSpace(part[idx+1:])
		switch k {
		case "v":
			version = v
			hasV = true
		case "kid":
			kid = v
		case "fp":
			fp = strings.ToLower(v)
			hasFp = true
		default:
			// Forward-compat: ignore unknown fields rather than reject.
		}
	}

	if !hasV {
		return nil, errors.New("DNS TXT record missing required 'v' field")
	}
	if version != "schemapin1" {
		return nil, fmt.Errorf("DNS TXT unsupported version: %s", version)
	}
	if !hasFp {
		return nil, errors.New("DNS TXT record missing required 'fp' field")
	}
	if !strings.HasPrefix(fp, "sha256:") {
		return nil, fmt.Errorf("DNS TXT 'fp' must be sha256:<hex>: %s", fp)
	}

	return &DnsTxtRecord{
		Version:     version,
		Kid:         kid,
		Fingerprint: fp,
	}, nil
}

// VerifyDnsMatch cross-checks the DNS TXT record's fingerprint against the
// discovery document's public key.
//
// Returns nil when the fingerprint matches the SHA-256 fingerprint of the
// public key in discovery.PublicKeyPEM. Returns an error otherwise. Callers
// MUST treat any non-nil return as a hard verification failure
// (verification.ErrDomainMismatch).
func VerifyDnsMatch(disc *discovery.WellKnownResponse, txt *DnsTxtRecord) error {
	if disc == nil {
		return errors.New("DNS TXT match requires a non-nil discovery document")
	}
	if txt == nil {
		return errors.New("DNS TXT match requires a non-nil record")
	}
	keyManager := crypto.NewKeyManager()
	computed, err := keyManager.CalculateKeyFingerprintFromPEM(disc.PublicKeyPEM)
	if err != nil {
		return fmt.Errorf("DNS TXT match: failed to compute fingerprint: %w", err)
	}
	computed = strings.ToLower(computed)
	if computed != txt.Fingerprint {
		return fmt.Errorf("DNS TXT fingerprint mismatch: discovery=%s, dns=%s", computed, txt.Fingerprint)
	}
	return nil
}

// TxtRecordName returns the DNS lookup name for a given tool domain
// ("_schemapin." prepended; trailing dot stripped).
func TxtRecordName(domain string) string {
	return "_schemapin." + strings.TrimSuffix(domain, ".")
}

// FetchDnsTxt fetches and parses the _schemapin.{domain} TXT record using
// the Go stdlib resolver.
//
// Returns:
//   - (record, nil) -- record present and parseable
//   - (nil, nil)    -- no _schemapin TXT record exists for the domain
//   - (nil, err)    -- DNS resolution error or the record is malformed
//
// Multi-chunk TXT records are concatenated per RFC 1464 (in emit order).
// When multiple TXT records exist at the same name, the first one whose
// joined value contains "v=schemapin1" is selected.
func FetchDnsTxt(ctx context.Context, domain string) (*DnsTxtRecord, error) {
	name := TxtRecordName(domain)
	resolver := &net.Resolver{}
	records, err := resolver.LookupTXT(ctx, name)
	if err != nil {
		var dnsErr *net.DNSError
		if errors.As(err, &dnsErr) && dnsErr.IsNotFound {
			return nil, nil
		}
		return nil, fmt.Errorf("DNS TXT lookup failed for %s: %w", name, err)
	}

	// Note: net.Resolver.LookupTXT already concatenates multi-chunk strings
	// within a single TXT RR into one entry, so each element here is the
	// full record.
	for _, rec := range records {
		if strings.Contains(rec, "v=schemapin1") {
			return ParseTxtRecord(rec)
		}
	}
	return nil, nil
}

File: python/schemapin/dns.py
"""DNS TXT cross-verification for SchemaPin v1.4.

A tool provider MAY publish a TXT record at ``_schemapin.{domain}`` containing
the public-key fingerprint advertised in ``.well-known/schemapin.json``. When
present, clients use it as a *second-channel* verification: the DNS
credential chain is independent of the HTTPS hosting credential chain, so
compromising one does not compromise the other.

TXT record format::

    _schemapin.example.com. IN TXT "v=schemapin1; kid=acme-2026-01; fp=sha256:a1b2c3..."

Fields:

- ``v`` -- version tag (``schemapin1``); required
- ``fp`` -- key fingerprint (``sha256:<hex>``); required, lowercase hex
- ``kid`` -- optional key id, used for disambiguating multi-key endpoints

Verification semantics:

- **Absent record** -- no effect (DNS TXT is optional)
- **Present and matching** -- confidence boost (no warning emitted; absence
  of mismatch is the signal)
- **Present and mismatching** -- hard failure with
  :class:`schemapin.verification.ErrorCode.DOMAIN_MISMATCH`

Use :func:`parse_txt_record` to parse a raw TXT string and
:func:`verify_dns_match` to cross-check it against a discovery document.
:func:`fetch_dns_txt` performs the DNS lookup; it requires the optional
``dnspython`` dependency (``pip install schemapin[dns]``).
"""

from dataclasses import dataclass
from typing import Any, Dict, Optional

from .crypto import KeyManager


@dataclass
class DnsTxtRecord:
    """Parsed ``_schemapin.{domain}`` TXT record.

    ``fingerprint`` is the lowercase string including the ``sha256:`` prefix
    so it can be compared directly with
    :meth:`KeyManager.calculate_key_fingerprint` output.
    """

    version: str
    fingerprint: str
    kid: Optional[str] = None


def parse_txt_record(value: str) -> DnsTxtRecord:
    """Parse a raw TXT record value.

    Example input::

        v=schemapin1; kid=acme-2026-01; fp=sha256:a1b2c3...

    Whitespace around ``;`` and ``=`` is tolerated. Field order is not
    significant. Unknown fields are ignored (forward-compat).

    Raises:
        ValueError: When the record is missing the required ``v`` or ``fp``
            fields, when the version is not ``schemapin1``, when ``fp`` lacks
            the ``sha256:`` prefix, or when a field is missing ``=``.
    """
    version: Optional[str] = None
    kid: Optional[str] = None
    fp: Optional[str] = None

    for raw_part in value.split(";"):
        part = raw_part.strip()
        if not part:
            continue
        if "=" not in part:
            raise ValueError(f"DNS TXT field missing '=': {part}")
        k, v = part.split("=", 1)
        k = k.strip().lower()
        v = v.strip()
        if k == "v":
            version = v
        elif k == "kid":
            kid = v
        elif k == "fp":
            fp = v.lower()
        # Forward-compat: ignore unknown fields rather than reject.

    if version is None:
        raise ValueError("DNS TXT record missing required 'v' field")
    if version != "schemapin1":
        raise ValueError(f"DNS TXT unsupported version: {version}")
    if fp is None:
        raise ValueError("DNS TXT record missing required 'fp' field")
    if not fp.startswith("sha256:"):
        raise ValueError(f"DNS TXT 'fp' must be sha256:<hex>: {fp}")

    return DnsTxtRecord(version=version, kid=kid, fingerprint=fp)


def verify_dns_match(
    discovery: Dict[str, Any], txt: DnsTxtRecord
) -> None:
    """Cross-check the DNS TXT record's fingerprint against the discovery doc.

    Computes the SHA-256 fingerprint of ``discovery["public_key_pem"]`` and
    compares it (case-insensitively) with ``txt.fingerprint``.

    Raises:
        ValueError: When the discovery document lacks ``public_key_pem`` or
            the fingerprints do not match. The mismatch case is what callers
            should map to ``ErrorCode.DOMAIN_MISMATCH``.
    """
    public_key_pem = discovery.get("public_key_pem")
    if not public_key_pem:
        raise ValueError(
            "Discovery document missing 'public_key_pem'; cannot compute fingerprint"
        )
    computed = KeyManager.calculate_key_fingerprint_from_pem(public_key_pem).lower()
    if computed != txt.fingerprint:
        raise ValueError(
            f"DNS TXT fingerprint mismatch: discovery={computed}, dns={txt.fingerprint}"
        )


def txt_record_name(domain: str) -> str:
    """Construct the DNS lookup name for a given tool domain."""
    return f"_schemapin.{domain.rstrip('.')}"


def fetch_dns_txt(domain: str) -> Optional[DnsTxtRecord]:
    """Fetch and parse the ``_schemapin.{domain}`` TXT record.

    Requires the optional ``dnspython`` dependency. Install with::

        pip install schemapin[dns]
        # or
        pip install dnspython

    Returns:
        The parsed record when present, or ``None`` when no
        ``_schemapin.{domain}`` TXT record exists.

    Multiple matching TXT chunks are joined per RFC 1464 (concatenation in
    emit order). Multiple separate TXT records at the same name are not
    supported -- the first record containing ``v=schemapin1`` wins.

    Raises:
        ImportError: When ``dnspython`` is not installed.
        ValueError: When the record exists but is malformed.
        RuntimeError: For DNS resolution errors other than NXDOMAIN/NoAnswer.
    """
    try:
        import dns.exception
        import dns.resolver
    except ImportError as e:
        raise ImportError(
            "fetch_dns_txt requires the 'dnspython' package; "
            "install schemapin[dns] or pip install dnspython"
        ) from e

    name = txt_record_name(domain)
    try:
        answer = dns.resolver.resolve(name, "TXT")
    except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):
        return None
    except dns.exception.DNSException as e:
        raise RuntimeError(f"DNS TXT lookup failed for {name}: {e}") from e

    for record in answer:
        # dnspython exposes TXT chunks via .strings (list of bytes) per RFC 1464.
        chunks = getattr(record, "strings", None) or []
        try:
            joined = "".join(
                c.decode("utf-8", errors="replace") if isinstance(c, (bytes, bytearray)) else str(c)
                for c in chunks
            )
        except Exception:
            joined = str(record)
        if "v=schemapin1" in joined:
            return parse_txt_record(joined)
    return None
